src/common/tree/tree_node.py

```diff
@@ -255,68 +255,45 @@ def __init__(
     def update(self, data: dict, partial_update: bool = False):
+        if partial_update:
+            self.entity.update(data)
+        else:
+            self.entity = data
+
         if data.get("_id"):
             self.set_uid(data.get("_id"))
 
-        # Set self.type from posted type, and validate against parent blueprint
+        # Set self.type from posted type
         self.type = data.get("type", self.attribute.attribute_type)
-        # Modify and add for each key in posted data
-        for key in data.keys():
-            new_data = data[key]
-            if key == "_id":
-                self.entity["_id"] = new_data
-                continue
-            attribute = self.blueprint.get_attribute_by_name(key)
-
-            if key == "type":
-                # Should always be able to specify type
-                self.entity[key] = new_data
 
-            # Add/Modify primitive data
-            if attribute.is_primitive:
-                self.entity[key] = new_data
-            # Add/Modify complex data
-            else:
-                child = self.get_by_path([key])
-                if not child:  # A new child has been added
-                    if attribute.is_array:
-                        child = ListNode(
-                            attribute.name,
-                            attribute,
-                            None,
-                            new_data,
-                            self,
-                            self.blueprint_provider,
-                            recipe_provider=self.recipe_provider,
-                        )
-                    else:
-                        child = Node(
-                            attribute.name,
-                            attribute,
-                            None,
-                            new_data,
-                            self,
-                            self.blueprint_provider,
-                            recipe_provider=self.recipe_provider,
-                        )
-                child.update(new_data, partial_update)
-
-        # Remove any child that is not specified in blueprint
-        # (Can happen if we change the type, and the old node had some children)
-        for child in self.children:
-            if child.attribute.name not in self.blueprint.get_attribute_names():
-                self.remove_by_path([child.key])
-
-        # Remove for every key in blueprint not in data or is a required attribute
-        removed_attributes = [attr for attr in self.blueprint.attributes if attr.name not in data]
-        for attribute in removed_attributes:
-            # Pop primitive data
-            if attribute.is_primitive:
-                self.entity.pop(attribute.name, None)  # type: ignore
-            # Remove complex data
+        self.children = []
+        complex_attributes = self.blueprint.get_none_primitive_types()
+        for attribute in complex_attributes:
+            if attribute.name not in self.entity:
```

Comment:
Should we add a comment here? I guess this can only happen if we change the type, and then we need to remove any attribute that is not specified in the blueprint.

Reply:
This check is for optional attributes.
We now build the list of children from scratch, meaning there is nothing to remove. And it will always be correct, no matter how much you change types